### PR TITLE
[Forwardport] Resolved : JS files located outside the web/js directory

### PR DESCRIPTION
**Pull request by hitesh-wagento**
### Description
Some JS files are direct children of `web` rather than `web/js`. This does not follow instructions from the [dev docs](https://devdocs.magento.com/guides/v2.2/javascript-dev-guide/javascript/custom_js.html) thus is confusing.

### Fixed Issues (if relevant)

1. <https://github.com/magento/magento2/issues/16302>: JS files located outside the web/js directory

### Expected result
1. JS files to be inside `web/js` as per the dev docs.

### Actual result
1. JS files are direct children of `web`, this is inconsistent with the dev docs.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### Original PR 
https://github.com/magento/magento2/pull/16582

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @hitesh-wagento. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

**Comment by hitesh-wagento** [comment]
Hi @VladimirZaets

As per you comment in https://github.com/magento/magento2/pull/16582 I have created new PR based on 2.3-develop branch.

**Comment by ihor-sviziev** [comment]
@VladimirZaets @hitesh-wagento what if we will just create map between old and new paths? In this case - this change will be backward compatible, so we can add them even to 2.2.x. Am I wrong?
Example:
```javascript
var config = {
    map: {
        '*': {
            transparent: 'Magento_Payment/js/transparent',
            'Magento_Payment/transparent': 'Magento_Payment/js/transparent'
        }
    }
};
```

**Comment by larsroettig** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @larsroettig, here is your [new Magento instance](https://pr-16708.engcom.dev.magento.com).
Admin access: [https://pr-16708.engcom.dev.magento.com/admin](https://pr-16708.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

**Comment by larsroettig** [comment]
Manual Testing 
- Create new Produkt Backend
- Checkout Produkt
- Register Account
- Captcha Passoword Forget
- Captcha Register
- Captcha Login 
- Order Grid
- Forget Password


Manual Testing looks very good

**Comment by hitesh-wagento** [comment]
@larsroettig 

Any update in this PR ?

Thanks

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
1. Please create PR to `2.2-develop` branch first

2. Please implement changes in all requirejs-config.js files that I described in https://github.com/magento/magento2/pull/16708#issuecomment-404262484 (in 2.2-develop PR only)

3. Please create separate commit for separate module in order to have ability apply these changes as described in https://support.magento.com/hc/en-us/articles/360005484154-Create-a-patch-for-a-Magento-2-Composer-installation-from-a-GitHub-commit

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev 

Previously I have created  PR based to 2.2-develop but @VladimirZaets told me created PR based on 2.3-develop. could you please suggest what can I do ?

https://github.com/magento/magento2/pull/16582#issuecomment-403016268

_______________
3. Please create separate commit for separate module in order to have ability apply these changes as described in https://support.magento.com/hc/en-us/articles/360005484154-Create-a-patch-for-a-Magento-2-Composer-installation-from-a-GitHub-commit
==> Means do you need to 27 (Specific Module requirejs-config.js or rename files) files commit in one branch but different commit ID ?

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento ,
1. These changes were requested to be done in 2.3-develop because they are not backward compatible. Now I added comment with info how we can make them backward compatible. So let's add backward compatible changes to 2.2-develop, 
2. I thought it will be few commits. Let's  **not** split it to commit per module

If you have any questions - feel free to contact me in slack

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
We are having discussion about this specific case with @VladimirZaets, so please wait once we will finalize out general vision how it should be done. I will update you after discussion.

**Comment by VladimirZaets** [comment]
Hi @hitesh-wagento. Please apply the fix provided by @ihor-sviziev to Magento 2.2 release line. 
For Magento 2.3 release line your fix is ok and will be processed after PR to Magento 2.2 will be merged.

**Comment by hitesh-wagento** [comment]
@VladimirZaets Thanks for reply.

**Comment by hitesh-wagento** [comment]
Hi @VladimirZaets 

I have completed all changes `2.2-develop` related https://github.com/magento/magento2/pull/16582

Thanks

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
I can't find changes that you did in: https://github.com/magento/magento2/pull/16582/files#diff-ce4bc7c513c965385f8696ebc9545f42
Everything else looks good

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev 

you can find changes in https://github.com/magento/magento2/pull/16708/commits/0ce439b1ecc2906dd840f37989c7c8ccf84dc6bf

Thanks

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
Sorry, but I can't see changes in send friend module.
Could you double-check it?

**Comment by hitesh-wagento** [comment]
Hello @ihor-sviziev 

I have checked again all module and added send friend module changes.

Thanks

**Review by ihor-sviziev — APPROVED** [review]

**Comment by hitesh-wagento** [comment]
@ihor-sviziev 

Any update in this PR ?

Thanks

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
Looks like we have conflict during merging 2.3-develop and your PR. Probably it caused by moving some files.

Could squash all your changes into single commit and force push your changes or merge 2.3-develop branch?

**Comment by hitesh-wagento** [comment]
Hello @ihor-sviziev 

Could you please share with me steps how to merge all commits in one commit?

Thanks

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento  could you just merge 2.3-develop branch from magento/magneto2 repository? It will be much easier

File: app/code/Magento/Captcha/view/frontend/web/js/onepage.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
```
Comment (ihor-sviziev) [inline app/code/Magento/Captcha/view/frontend/web/js/onepage.js]: This file was removed in https://github.com/magento/magento2/commit/efdc2e39a07828c505611ceb8f537ee35eb8aeae#diff-90e2539e902c667160d28b2cb3c56486
No need to create it again.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Captcha/view/frontend/web/js/onepage.js (hunk 1 above)
Comment (hitesh-wagento) [reply app/code/Magento/Captcha/view/frontend/web/js/onepage.js]: Removed and push again

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-2895](https://jira.corp.magento.com/browse/ENGCOM-2895) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
Hi @hitesh-wagento. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.0. 
Please check the release notes for final confirmation.

**Pull request merged**